### ISSUE TITLE
Document Possible Issues with Apple M1 go Install

### ISSUE THREAD

**Issue by andorsk** [Dentity-Development/dentity-ecosystems-sdk#1142]
#### Hardware Overview:
  Model Name:	MacBook Air
  Model Identifier:	MacBookAir10,1
  Chip:	Apple M1
  Total Number of Cores:	8 (4 performance and 4 efficiency)
  Memory:	8 GB
  System Firmware Version:	7459.121.3
  OS Loader Version:	7459.121.3

### OS 
Monterey 12.4 
 
### Steps

1. Install okapi over brew
2. Run the test with : LD_LIBRARY_PATH=/opt/homebrew/lib go test ./... -run TestVaccineDemo

### Reproduction Code

Section of code. Note: I pulled this out from a larger piece of code and haven't tested this particular segment alone.

```

func makeTestEcosystem(tt *testing.T) (*services.Trinsic, *provider.CreateEcosystemResponse) {
	trinsic, err := services.NewTrinsic(services.WithTestEnv())
	assert.NoError(tt, err)

	provider := trinsic.Provider()
	assert.NoError(tt, err)

	eco, err := provider.CreateEcosystem(context.Background(), nil)
	assert.NoError(tt, err)

	return trinsic, eco
}

func createActors(t *testing.T, trinsic *services.Trinsic, ecoId string) map[string]string {
	actors := []string{allison, airline, clinic}
	ret := make(map[string]string)
	for _, a := range actors {
		actor, err := trinsic.Account().LoginAnonymous(context.Background(), ecoId)
		assert.NoError(t, err)
		ret[a] = actor
	}
	return ret
}

// Test here
func TestVaccineDemo(t *testing.T){
	var err error
	trinsic, eco := makeTestEcosystem(t)
	assert.NotNil(t, trinsic)

	actors := createActors(t, trinsic, eco.Ecosystem.Id)
	assert.Len(t, 3, actors)

	trinsic.SetAuthToken(actors[clinic])
	info, err := trinsic.Account().GetInfo(context.Background())
}
```

`LD_LIBRARY_PATH=/opt/homebrew/lib go test ./... -run TestVaccineDemo`

```
panic: unable to open a handle to the library [recovered]
        panic: unable to open a handle to the library

goroutine 8 [running]:
testing.tRunner.func1.2({0x4778200, 0xc0001acee0})
        /usr/local/go/src/testing/testing.go:1396 +0x24e
testing.tRunner.func1()
        /usr/local/go/src/testing/testing.go:1399 +0x39f
panic({0x4778200, 0xc0001acee0})
        /usr/local/go/src/runtime/panic.go:884 +0x212
github.com/trinsic-id/okapi/go/okapi.getFunctionPointer({0x487cce6, 0xe})
        /Users/akmb2/go/pkg/mod/github.com/trinsic-id/okapi/go@v1.6.0/okapi/native_call.go:81 +0x9d
github.com/trinsic-id/okapi/go/okapi.callOkapiNative({0x4a59fe0?, 0xc0000b1320?}, {0x4a5a000, 0xc000474820}, {0x487cce6, 0xe})
        /Users/akmb2/go/pkg/mod/github.com/trinsic-id/okapi/go@v1.6.0/okapi/native_call.go:47 +0x152
github.com/trinsic-id/okapi/go/okapi.okapiMetadata.GetMetadata({})
        /Users/akmb2/go/pkg/mod/github.com/trinsic-id/okapi/go@v1.6.0/okapi/metadata.go:20 +0x65
github.com/trinsic-id/sdk/go/services.(*serviceBase).BuildMetadata(0xc00010bf50, {0x4a5a360, 0xc0000b12c0})
        /Users/akmb2/go/pkg/mod/github.com/trinsic-id/sdk/go@v1.8.0/services/service_base.go:120 +0xbf
github.com/trinsic-id/sdk/go/services.(*serviceBase).GetMetadataContext(0xc00009bba0?, {0x4a61e90, 0xc00003e0c8}, {0x4a5a360?, 0xc0000b12c0?})
        /Users/akmb2/go/pkg/mod/github.com/trinsic-id/sdk/go@v1.8.0/services/service_base.go:104 +0x4d
github.com/trinsic-id/sdk/go/services.(*accountBase).GetInfo(0xc00007c300, {0x4a61e90, 0xc00003e0c8})        /Users/akmb2/go/pkg/mod/github.com/trinsic-id/sdk/go@v1.8.0/services/account_service.go:253 +0x62
github.com/benri-io/sdk/operators.TestVaccineDemo(0x0?)
        /Users/akmb2/workspace/github.com/benri-io/sdk/operators/trinsic_test.go:107 +0x12f
testing.tRunner(0xc00009bba0, 0x497f310)
        /usr/local/go/src/testing/testing.go:1446 +0x10b
created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:1493 +0x35f
```

**Comment by tmarkovski** [comment]
Interesting, I'm also on a M1 mac, though I'm on ventura. Mind checking that the libraries have the correct architecture?

```
lipo -info /opt/homebrew/lib/libokapi.dylib
lipo -info /opt/homebrew/lib/libokapi.a
```

You should get something like:

```
Non-fat file: /opt/homebrew/lib/libokapi.dylib is architecture: arm64
```

**Comment by fundthmcalculus** [comment]
Can confirm I see the error on my M1 machine, working on a fix...

**Comment by fundthmcalculus** [comment]
Developing the auto loading capability to look in the normal `brew install` directories appears to fix the issue. Update to go SDK v1.8.1. I got your sample working here: 
https://github.com/trinsic-id/sdk-examples/pull/169

**Comment by andorsk** [comment]
awesome! thanks @fundthmcalculus

**Comment by andorsk** [comment]
Still not working for me unfortunately:

I think the fix you applied is helpful, b/c it probably removes the need for me to update LD_LIBRARY_PATH during runtime w/ the command: `LD_LIBRARY_PATH=/opt/homebrew/lib go test ./... -run TestVaccineDemo`

but I still hit the same error: 

go.sum
```
github.com/trinsic-id/okapi/go v1.6.1 h1:mdp/ixgCgV+E/ptZQBpwE3iK3ZIl+ynQpnxMxB3c58k=
github.com/trinsic-id/okapi/go v1.6.1/go.mod h1:Is/+ttfc/M1XS+84JTztm7oqV0vrqwzA6qyuufIs5Cg=
github.com/trinsic-id/sdk/go v1.8.1 h1:yhF13/hTyaQ6zINSJpYYqm0AxGMXhcJPHgaUp/MfLTo=
github.com/trinsic-id/sdk/go v1.8.1/go.mod h1:BRhiB8lIZ4esmcHJ0G+XkFQQp3tYX1ZFE1SQSeJOOZI=
```

Libs: 
```
➜  operators git:(credential) ✗ lipo -info /opt/homebrew/lib/libokapi.a
Non-fat file: /opt/homebrew/lib/libokapi.a is architecture: arm64
➜  operators git:(credential) ✗ lipo -info /opt/homebrew/lib/libokapi.dylib
Non-fat file: /opt/homebrew/lib/libokapi.dylib is architecture: arm64
```

**Comment by andorsk** [comment]
I reinstalled okapi just in case: `arch -arm64 brew install okapi` and copied the sdk-examples sample. Same thing.

**Comment by fundthmcalculus** [comment]
Okay this is a little weird. The error message is not spitting out a list of paths searched which means it found the proper Library file. Now it makes me think that there's some security setting on Mac OS which is blocking access to the executable file. Unfortunately, this gets out of my zone of expertise because I don't use mac.

**Comment by andorsk** [comment]
This is very possible actually.

**Comment by andorsk** [comment]
I think I got it! https://stackoverflow.com/questions/70482886/how-to-compile-an-amd64-binary-that-uses-c-on-an-m1-arm64-mac led me to do this:

```
CGO_ENABLED=1 GOOS=darwin GOARCH=arm64  go test ./...
```

I think probably CGO_ENABLED was not set and that was causing the linking error.

**Comment by fundthmcalculus** [comment]
You don't need `LD_LIBRARY_PATH`. Because MacOS typically blocks (makes empty) that environment variable, we use `OKAPI_LIBRARY_PATH` for cases where you need to manually load the library.

Check into the security settings and let me know. I should be available in a couple hrs if you need me.

**Comment by fundthmcalculus** [comment]
Ah, okay!!! Awesome!!!!

**Comment by andorsk** [comment]
I'll close this once I've done a little more testing. Thanks @fundthmcalculus for the support here!  Might be worth to put something in the docs about this, but let me do a little more digging and make sure I understand specifically what flag omission was breaking things. 

Might make sense to put a Makefile or something for people to test against darwin, idk.

**Comment by fundthmcalculus** [comment]
Definitely agree we should document this. Do you know if there's an easy way to throw a compile error if CGO is not enabled? (Or, better yet, just enable it automatically)

We can definitely add a makefile to the example if that would be helpful!

Ironically, the OS with the least issues on dynamic loading native binaries is...:drumroll: Windows. LOL.

**Comment by fundthmcalculus** [comment]
Don't close this, just comment if everything works for you. I'll hijack this ticket for the documentation/makefile work.

**Comment by andorsk** [comment]
OK. I did a little more debugging. It's actually not CGO ( my bad ) but GOARCH not being set to arm64 that was causing the error. CGO also needs to be set, but mine was actually already enabled by default. 

I think GOARCH defaults to `amd64`?

```
➜  operators git:(credential) ✗ go env GOOS GOARCH
darwin
amd64
```

Because these are not explicitly set by me to be amd64:
```
➜  operators git:(credential) ✗ echo "os: $GOOS arch: $GOARCH"
os:  arch: 
``` 

FYI, when CGO_ENABLED is not set it gives a series of errors like this: 
```
/Users/akmb2/go/pkg/mod/github.com/trinsic-id/okapi/go@v1.6.1/okapi/didcomm.go:22:9: undefined: callOkapiNative
```

Whereas the GOARCH is incorrect, it gives the `panic: unable to open a handle to the library error.`

**Comment by andorsk** [comment]
So conclusion here: `GOARCH` might be set incorrect on the `go env`, which will cause the `unable to open a handle` error. Setting `GOARCH` to be correct with the actual architecture is the fix.

**Comment by andorsk** [comment]
I'll just mention one last thing here:  It might seem like 'duh' of course GOARCH needs to be aligned with your system, but;

1) I had a bad assumption my GOARCH was correct already. After all, other tests were working fine and I didn't set it on my own. 2) The messaging doesn't necessarily indicate clearly that somehow GOARCH was not right.

So, GOARCH == ARCH you're running on is obvious, but could be a gotcha if somehow there's a misalignment that people waste cycles on, so the fix from the libs perspective is to give some better feedback/documentation so it's easier to catch the alignment issue I think. 

I'm not sure what that looks like though. I think having this issue documented on the issues is really a solid first step though.

**Comment by andorsk** [comment]
So I think I found out how this happened. Turns out that my entire Go install was done on the wrong arch and I had no idea, so this was actually really good to catch that I had a bigger issue on my computer.

If you're installing go on M1, the [Standard Go.dev Install Page](https://go.dev/doc/install) actually defaults to still showing you an amd64 install pkg even if you're on an M1. My guess is somehow I installed the amd64 version, but with rosetta powering it, it was still working until I started looking into the linker issue. So I never noticed there was a problem with my install until now. You could be running the amd go version I think for a while without ever catching it. 

I had to go to [other downloads](https://go.dev/dl/) to get arm. I must have missed it before b/c for some reason I had assumed that just like it detected the correct OS, it would have detected the correct arch during install. 

The long story short is though, when that happens, it actually sets all the environment variables also to AMD, including `GOHOSTARCH`, `GOARCH`, and `GOTOOLDIR`. This, then causes the bad env variable to impact the linker. 

I don't think the scope of the libs is to make sure that you have installed Go correctly. I'm OK saying that was me having a stupid issue with my environment. I think having this issue as a reference, in case someone makes a dumbo mistake like I did, is probably sufficient.
